### ISSUE TITLE
Support for copying multiple files

### ISSUE THREAD

**Issue by Lightjohn** [crazyscot/qcp#21]
**Describe the solution you'd like**
This is the same as what you say in the readme :)
Just emulate scp -r in qcp or copy multiple files / folders

**Additional context**
Just creating an issue so that anyone interested in that feature can just follow the progress with one issue 

Just close it if you don't like the idea

Great work ;)

**Comment by terrorbyte** [comment]
This has been one of the consistent things I've wanted (and slowly am working on) as well. From what I can tell it requires:
- Updating the schema to use a `fs::Metadata` style list (might depend on #12 since continuing to perpetuate that if it's going away might not be ideal). That also requires some thinking about how to handle permission mapping/symlink footguns.
- Changing the progress bar to be the aggregate size
- Checks for file changes mid or pre-upload
- How to fit the multiple HTTP/2 streams into the new multi-file structure. ie does qcp just copy one file at a time or split up chunks or prioritize etc.

**Comment by crazyscot** [comment]
I've had multiple files in mind since day one, just a question of finding the time to do it :smile: 

My initial thoughts were to use the existing session protocol as far as possible; open multiple streams, one per file. Obviously to support Get we'd need to add some sort of list-files command that supported globs and/or recursion. (Thereby making the protocol more of a generic file transfer protocol.) There are undoubtedly other ways to do it which can be explored, but that would at least be a way to get going quickly.

I hadn't thought about file changes whilst transfer was in flight. To date I was working off the standard model in that we'd hold a reader-lock throughout the transfer, so it could not be modified (though of course it could be unlinked and another file of the same name created). It seems a bit... _specialised_? (To be clear, I'm not opposed to supporting it, it's just that I don't have a use case.)

If we were to ever split up a file transfer into chunks I suspect we'd want to include some sort of hash check for integrity, but that's perhaps later down the track? For a package that's still in its 0.1 series (0.2 coming soon when I finish configuration file support) I consider iterating quickly to be more important than getting features perfect first time!

**Comment by crazyscot** [comment]
Capturing my implementation ideas to date...

Multi file and recursive transfer are very similar operations, so we might be able to implement one set of protocol changes covering both.

Parallel file transfer is likely to be important for efficiency, but there's an interesting question: should the streams be prioritised?

## CLI

### Multi file transfer

On `Get` the CLI sees multiple names, or a shell glob which the server expands. (It's up to the user to quote a shell glob; we can't save them from their shell.)

On `Put` the CLI most likely sees the result of expanding a shell glob i.e. multiple file arguments.

### Recursive transfer

To `Get` recursively, you specify a single directory. The server expands.

To `Put` recursively is similar, the client expands into multiple individual files.

## Protocol ideas

### Put

* This is a number of Put commands, but potentially in parallel.
* In order to create files recursively we must be prepared to create directories. This maybe mean a new command `EnsureDirectoryExists(path)`, which of course has to be completed before we attempt to Put any files within that directory. 
  * Parallelism may lead to some interesting edge cases here!
  * Perhaps it's `EnsureDirectoriesExist(array of paths)`.
  * Perhaps creating the parent directory should in fact be implicit in Put, or at least a flag in Put that means "mkdir -p first"?
* We will need to think carefully about permissions when creating directories.

### Get

We'll need to add some sort of `GetMulti` command to the session protocol. Options identified to date:

1.  `GetMulti(Spec)` returns a list of files. Client retains control, it sends multiple Gets.
2. `GetMulti(Spec)` causes the server to get on with it and open multiple file streams.

a `GetMultiSpec` consists of:
* Path
* ExpandGlob flag
* Recurse flag

Edge cases to think about:
* symlinks
* hardlinks
* filenames containing characters that would ordinarily be glob-expanded

## Configuration

* Parallel stream limit. This would be negotiated as `min(server, client)`. The idea is that either side might need to limit disk I/O.

## Styling changes

The existing CLI might need a little work to effectively display multiple files at once.

**Comment by terrorbyte** [comment]
Another couple of things to consider:

* Failure cases for multiple files - Does a file/directory creation or write error count as a fatal error for all qcp or is it soft, error but continue. could be a toggle flag
* Resumption - might be a bit out of scope but some similar file copying things either create checkpoint files with known state or generate their canonical list in a way that that could be passed as a tokenized flag for resumption on failures (I have seen one implementation that represents the list of canonical files as a idempotent merkle tree and then if a failure happens the resumption will print the leaf where it happens, so when resumption happens it can just traverse the tree to identify where the failure was removing the need to keep full state somewhere).
* TOCTOU issues (partially covered by the symlink conversation) - but just ensuring that file creation with expected permissions happens before the write, ie not creating the file first and then assuming the permissions.

Stoked for this feature.

**Comment by crazyscot** [comment]
Now I've got the test/refactoring out of the way the path to implement this is clearing. I've tentatively pinned 0.6 on it, though I think I want to get at least basic file integrity checking in place first.

**Comment by crazyscot** [comment]
State of play:
- [x] Local source (PUT) multi files - manually specified: Merged.
- [x] Local source (PUT) CLI multi files - glob expansion: Your shell likely does this, so we get it for free?
- [x] Remote source (GET) multi files - manually specified: Merged.
- [ ] Remote source (GET) multi files - server expands glob
- [ ] Remote source (GET) directory recursion
- [ ] Local source (PUT) directory recursion

**Comment by crazyscot** [comment]
> * How to fit the multiple HTTP/2 streams into the new multi-file structure. ie does qcp just copy one file at a time or split up chunks or prioritize etc.

At the moment we open one stream at a time. Obviously this isn't very performant, and moving to 2 or 3 streams in parallel (#184) will help a lot.
